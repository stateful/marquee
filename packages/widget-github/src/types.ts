import type { ContextProperties } from '@vscode-marquee/utils'

export interface Since {
  name: string
  value: string
}

export interface Contributor {
  username: string
  avatar: string
}

export interface SpokenLanguage {
  name: string
  urlParam: string
}

export interface Trend {
  description: string
  url: string
  author: string
  name: string
  currentPeriodStars: number
  language: string
  languageColor: string
  forks: number
  stars: number
  builtBy: Contributor[]
}

export type SinceConfiguration = 'Daily' | 'Weekly' | 'Monthly'
export interface Configuration {
  since?: SinceConfiguration
  language?: string
  spoken?: string
  trendFilter?: string
}

export interface State {}

export interface ContextTypes extends Configuration {
  trends: Trend[]
  isFetching: boolean
  showDialog: boolean
  error?: Error
}

export interface Context extends ContextProperties<ContextTypes> {
  _updateSpoken: (id: SpokenLanguage) => void
  _updateLanguage: (val: SpokenLanguage) => void
  _updateSince: (val: Since) => void
  _updateFilter: (trendFilter: string) => void
}

export interface Events {
  openGitHubDialog: boolean
}
