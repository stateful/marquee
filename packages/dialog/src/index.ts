import DialogContainer from './DialogContainer'
import DialogTitle from './DialogTitle'

export { DialogContainer, DialogTitle }
