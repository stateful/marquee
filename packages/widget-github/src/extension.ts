import vscode from 'vscode';

import ExtensionManager from '@vscode-marquee/utils/extension';

import { DEFAULT_CONFIGURATION, DEFAULT_STATE } from './constants';
import type { Configuration } from './types';

const STATE_KEY = 'widgets.github';

export function activate (
  context: vscode.ExtensionContext,
  channel: vscode.OutputChannel
) {
  const stateManager = new ExtensionManager<{}, Configuration>(context, channel, STATE_KEY, DEFAULT_CONFIGURATION, DEFAULT_STATE);

  return {
    marquee: {
      disposable: stateManager,
      defaultState: stateManager.state,
      defaultConfiguration: stateManager.configuration,
      setup: stateManager.setBroadcaster.bind(stateManager)
    }
  };
}
