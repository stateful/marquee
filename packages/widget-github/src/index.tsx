import React from 'react';
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faFire } from "@fortawesome/free-solid-svg-icons/faFire";

import Github from "./Widget";

export default {
  name: "github",
  icon: <FontAwesomeIcon icon={faFire} />,
  tags: ["github", "trending", "projects", "entertainment"],
  label: 'GitHub',
  description: "The Github trending feed.",
  component: Github,
};
