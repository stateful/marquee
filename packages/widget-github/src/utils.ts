import type { MarqueeWindow } from '@vscode-marquee/utils';

import { spokenLanguages, trendLanguages } from './constants';
import type { SinceConfiguration } from './types';

declare const window: MarqueeWindow;
const ERROR_MESSAGE = "Couldn't fetch GitHub trends!";

export async function fetchData (since?: SinceConfiguration, language?: string, spoken?: string) {
  const searchParams = new URLSearchParams({ props: window.marqueeUserProps });

  if (since) {
    searchParams.append('since', since.toLocaleLowerCase());
  }

  const languageParam = trendLanguages.find((l) => l.name === language);
  if (languageParam) {
    searchParams.append('language', languageParam.urlParam);
  }

  const spokenParam = spokenLanguages.find((l) => l.name === spoken);
  if (spokenParam) {
    searchParams.append('spoken', spokenParam.urlParam);
  }

  const url = `${window.marqueeBackendBaseUrl}/getRepositories?${searchParams.toString()}`;
  console.log(`Fetch repositories via: ${url}`);
  const res = await fetch(url).catch(
    () => { throw new Error(ERROR_MESSAGE); });

  if (!res.ok) {
    throw new Error(`${ERROR_MESSAGE} (status: ${res.status})`);
  }

  return res.json();
}
