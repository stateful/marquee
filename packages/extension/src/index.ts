import type vscode from "vscode";
import type { Client } from 'tangle';

import telemetry from './telemetry';
import { MarqueeExtension } from "./extension";
import { getExtProps } from './utils';

export function activate(context: vscode.ExtensionContext) {
  telemetry.sendTelemetryEvent('extensionActivate', getExtProps());
  new MarqueeExtension(context);

  if (process.env.NODE_ENV !== 'development') {
    return {};
  }

  return {
    /**
     * demo code for example widget in "/example"
     */
    marquee: {
      setup: (tangle: Client<{ counter: number }>) => {
        let i = 0;
        setInterval(() => {
          tangle.emit('counter', ++i);
        }, 1000);
      }
    }
  };
}

// this method is called when your extension is deactivated
export function deactivate() {
  telemetry.sendTelemetryEvent('extensionDeactivate');
}
