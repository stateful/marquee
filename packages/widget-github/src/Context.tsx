import React, { createContext, useState, useEffect } from "react";
import { connect, getEventListener, MarqueeEvents } from "@vscode-marquee/utils";

import { fetchData } from './utils';
import { DEFAULT_CONFIGURATION } from './constants';
import type { Context, Trend, Configuration, Since, SpokenLanguage, SinceConfiguration } from './types';

const TrendContext = createContext<Context>({} as any);

interface Props {
  children?: React.ReactNode;
}

const TrendProvider = ({ children }: Props) => {
  const widgetState = getEventListener<Configuration>('@vscode-marquee/github-widget');
  const providerValues = connect<Configuration>(DEFAULT_CONFIGURATION, widgetState);

  const [error, setError] = useState<Error>();
  const [isFetching, setIsFetching] = useState(false);
  const [trends, setTrends] = useState<Trend[]>([]);
  const [showDialog, setShowDialog] = useState(false);

  const _updateFilter = (trendFilter: string) => {
    providerValues.setTrendFilter(trendFilter);
  };

  const _updateSince = (since?: Since) => {
    providerValues.setSince(since?.name as SinceConfiguration || '');
  };

  const _updateLanguage = (language?: SpokenLanguage) => {
    providerValues.setLanguage(language?.name || '');
  };

  const _updateSpoken = (spoken?: SpokenLanguage) => {
    providerValues.setSpoken(spoken?.name || '');
  };

  useEffect(() => {
    const eventListener = getEventListener<MarqueeEvents>();
    eventListener.on('openGitHubDialog', setShowDialog);
  }, []);

  useEffect(() => {
    /**
     * don't fetch if
     */
    if (
      /**
       * we haven't received all configuration params
       */
      typeof providerValues.since !== 'string' ||
      typeof providerValues.language !== 'string' ||
      typeof providerValues.spoken !== 'string' ||
      /**
       * we are already fetching something
       */
      isFetching
    ) {
      return;
    }

    setIsFetching(true);
    fetchData(providerValues.since, providerValues.language, providerValues.spoken).then(
      (res) => {
        setTrends(res as Trend[]);
        setError(undefined);
      },
      (e: Error) => setError(e)
    ).finally(() => setIsFetching(false));
  }, [providerValues.language, providerValues.since, providerValues.spoken]);

  return (
    <TrendContext.Provider
      value={{
        ...providerValues,
        trends,
        error,
        isFetching,
        showDialog,
        setError,
        setTrends,
        setIsFetching,
        setShowDialog,
        _updateSpoken,
        _updateFilter,
        _updateSince,
        _updateLanguage
      }}
    >
      {children}
    </TrendContext.Provider>
  );
};

export default TrendContext;
export { TrendProvider };
