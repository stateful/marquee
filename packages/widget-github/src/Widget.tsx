import React, { useContext, useMemo } from 'react'
import { Grid, Link, Typography, Chip, Avatar, CircularProgress, Box } from '@mui/material'
import AvatarGroup from '@mui/material/AvatarGroup'
import wrapper, { Dragger, HidePop } from '@vscode-marquee/widget'
import StarIcon from '@mui/icons-material/Star'
import StarHalfIcon from '@mui/icons-material/StarHalf'
import FiberManualRecordIcon from '@mui/icons-material/FiberManualRecord'

import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { faCodeBranch } from '@fortawesome/free-solid-svg-icons/faCodeBranch'

import { NetworkError } from '@vscode-marquee/utils'

import TrendContext, { TrendProvider } from './Context'
import TrendingDialogLauncher from './components/TrendingDialog'
import Filter from './components/Filter'

let GChip = ({ ...rest }) => {
  return (
    <Chip
      variant="outlined"
      style={{ fontSize: '12px', border: 0 }}
      size="small"
      {...rest}
    />
  )
}

let Github = () => {
  const { trends, isFetching, error, trendFilter } = useContext(TrendContext)

  const filteredTrends = useMemo(() => {
    let filteredTrends = trends

    if (trendFilter) {
      let filteredArr = filteredTrends.filter((entry) => {
        return (
          entry.description.toLowerCase().indexOf(trendFilter.toLowerCase()) !==
          -1
        )
      })
      filteredTrends = filteredArr
    }

    return filteredTrends
  }, [trends, trendFilter])

  return (
    <>
      <Grid item xs={1} style={{ maxWidth: '100%' }}>
        <Box sx={{
          borderBottom: '1px solid var(--vscode-editorGroup-border)',
          padding: '8px',
        }}>
          <Grid
            container
            direction="row"
            wrap="nowrap"
            alignItems="stretch"
            alignContent="stretch"
            justifyContent="space-between"
          >
            <Grid item>
              <Typography variant="subtitle1">Trending on Github</Typography>
            </Grid>
            <Grid item>
              <Grid container direction="row" spacing={1}>
                <Grid item>
                  <Filter />
                </Grid>
                <Grid item>
                  <TrendingDialogLauncher />
                </Grid>
                <Grid item>
                  <HidePop name="github" />
                </Grid>
                <Grid item>
                  <Dragger />
                </Grid>
              </Grid>
            </Grid>
          </Grid>
        </Box>
      </Grid>
      <Grid item xs>
        <Grid
          container
          wrap="nowrap"
          direction="column"
          style={{ height: '100%' }}
        >
          <Grid item xs style={{ overflow: 'auto' }}>
            {error && (
              <Grid
                item
                xs
                style={{
                  overflow: 'auto',
                  height: '100%',
                  width: '100%',
                  padding: '24px',
                }}
              >
                <NetworkError message={error.message} />
              </Grid>
            )}
            {isFetching && (
              <Grid
                container
                style={{ height: '100%' }}
                alignItems="center"
                justifyContent="center"
                direction="column"
              >
                <Grid item>
                  <CircularProgress color="secondary" />
                </Grid>
              </Grid>
            )}
            {!isFetching && !error && filteredTrends.length === 0 && (
              <Grid
                container
                style={{ height: '100%' }}
                alignItems="center"
                justifyContent="center"
                direction="column"
              >
                <Grid item>
                  <Typography variant="body2">
                    There are no matches for your search criteria.
                  </Typography>
                </Grid>
              </Grid>
            )}
            {!isFetching && !error && filteredTrends.length !== 0 &&
              filteredTrends.map((entry) => {
                return (
                  <Grid
                    aria-labelledby="trend-entry"
                    key={entry.url}
                    container
                    direction="column"
                    sx={{
                      marginTop: '4px',
                      marginBottom: '4px',
                      padding: '16px',
                      borderBottom: '1px solid var(--vscode-editorGroup-border)',
                    }}
                  >
                    <Grid
                      container
                      direction="row"
                      justifyContent="space-between"
                      alignItems="center"
                    >
                      <Grid item>
                        <Link
                          component="a"
                          href={entry.url}
                          target="_blank"
                          underline="hover"
                        >
                          <Typography variant="body2">
                            {entry.author}/{entry.name}
                          </Typography>
                        </Link>
                      </Grid>
                      <Grid item>
                        <GChip
                          label={entry.currentPeriodStars.toLocaleString()}
                          icon={<StarHalfIcon />}
                        />
                      </Grid>
                    </Grid>
                    <Grid
                      container
                      direction="row"
                      justifyContent="space-between"
                      alignItems="center"
                    >
                      <Grid item>
                        <Typography variant="caption">
                          {entry.description}
                        </Typography>
                      </Grid>
                    </Grid>
                    <Grid container>
                      <Grid item>&nbsp;</Grid>
                    </Grid>
                    <Grid
                      container
                      direction="row"
                      justifyContent="space-between"
                      alignItems="flex-end"
                    >
                      <Grid item>
                        <Grid container spacing={1}>
                          {entry.language && (
                            <Grid aria-label="language" item>
                              <GChip
                                label={entry.language}
                                icon={
                                  <FiberManualRecordIcon
                                    style={{
                                      fill: `${entry.languageColor}`,
                                    }}
                                  />
                                }
                              />
                            </Grid>
                          )}
                          <Grid aria-label="forks" item>
                            <GChip
                              label={entry.forks.toLocaleString()}
                              icon={<FontAwesomeIcon icon={faCodeBranch} />}
                            />
                          </Grid>
                          <Grid aria-label="stars" item>
                            <GChip
                              label={entry.stars.toLocaleString()}
                              icon={<StarIcon />}
                            />
                          </Grid>
                        </Grid>
                      </Grid>
                      <Grid item>
                        <AvatarGroup>
                          {entry.builtBy.map((contributor) => {
                            return (
                              <Avatar
                                key={contributor.username}
                                style={{
                                  height: '22px',
                                  width: '22px',
                                  border: 0,
                                }}
                                src={contributor.avatar}
                                alt={contributor.username}
                              />
                            )
                          })}
                        </AvatarGroup>
                      </Grid>
                    </Grid>
                  </Grid>
                )
              })}
          </Grid>
        </Grid>
      </Grid>
    </>
  )
}

const Widget = () => (
  <TrendProvider>
    <Github />
  </TrendProvider>
)
export default wrapper(Widget, 'github')
