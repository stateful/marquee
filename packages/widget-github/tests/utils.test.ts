import { fetchData } from '../src/utils'

declare const window: {
  fetch: Function
  marqueeUserProps: any
  marqueeBackendBaseUrl: string
}

const fetchOrig = window.fetch
beforeEach(() => {
  window.fetch = jest.fn()
})

test('fetchData', async () => {
  window.marqueeUserProps = JSON.stringify({ foo: 'bar' })
  window.marqueeBackendBaseUrl = 'http://marquee.com';
  (window.fetch as jest.Mock).mockResolvedValue({
    ok: 1,
    json: () => 'foobar'
  })

  expect(await fetchData(
    'Daily',
    'TypeScript',
    'English'
  )).toBe('foobar')
  expect((window.fetch as jest.Mock).mock.calls).toMatchSnapshot()
})

afterAll(() => {
  window.fetch = fetchOrig
})
