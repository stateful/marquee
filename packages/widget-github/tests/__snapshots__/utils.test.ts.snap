// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetchData 1`] = `
Array [
  Array [
    "http://marquee.com/getRepositories?since=daily&language=typescript&spoken=en",
  ],
]
`;
